ArmPkg/Library/ArmFfaLib/ArmFfaCommon.c

```diff
@@ -0,0 +1,713 @@
+}
+
+/**
+  Get partition id.
+
+  @param [out]    PartId      Partition id
+
+  @retval EFI_SUCCESS
+  @retval Others              Error
+
+**/
+EFI_STATUS
+EFIAPI
+ArmFfaLibPartitionIdGet (
+  OUT UINT16  *PartId
+  )
+{
+  EFI_STATUS    Status;
+  ARM_FFA_ARGS  FfaArgs;
+
+  if (PartId == NULL) {
+    return EFI_INVALID_PARAMETER;
+  }
+
+  ZeroMem (&FfaArgs, sizeof (ARM_FFA_ARGS));
+
+  FfaArgs.Arg0 = ARM_FID_FFA_ID_GET;
+
+  ArmCallFfa (&FfaArgs);
+
+  if (IsFfaError (&FfaArgs)) {
+    Status = FfaStatusToEfiStatus (FfaArgs.Arg2);
+    DEBUG ((
+      DEBUG_ERROR,
+      "%a: Failed to get partition id. Status: %r\n",
+      __func__,
+      Status
+      ));
+    return Status;
+  }
+
+  *PartId = (FfaArgs.Arg2 >> ARM_FFA_DEST_EP_SHIFT) & ARM_FFA_PARTITION_ID_MASK;
+
+  return EFI_SUCCESS;
+}
+
+/**
+  Get spmc or spmd partition id.
+
+  @param [out]    SpmPartId      spmc/spmd partition id
+
+  @retval EFI_SUCCESS
+  @retval Others              Error
+
+**/
+EFI_STATUS
+EFIAPI
+ArmFfaLibSpmIdGet (
+  OUT UINT16  *SpmPartId
+  )
+{
+  EFI_STATUS    Status;
+  ARM_FFA_ARGS  FfaArgs;
+
+  if (SpmPartId == NULL) {
+    return EFI_INVALID_PARAMETER;
+  }
+
+  ZeroMem (&FfaArgs, sizeof (ARM_FFA_ARGS));
+
+  FfaArgs.Arg0 = ARM_FID_FFA_SPM_ID_GET;
+
+  ArmCallFfa (&FfaArgs);
+
+  if (IsFfaError (&FfaArgs)) {
+    Status = FfaStatusToEfiStatus (FfaArgs.Arg2);
+    DEBUG ((
+      DEBUG_ERROR,
+      "%a: Failed to get partition id. Status: %r\n",
+      __func__,
+      Status
+      ));
+    return Status;
+  }
+
+  *SpmPartId = (FfaArgs.Arg2 >> ARM_FFA_DEST_EP_SHIFT) & ARM_FFA_PARTITION_ID_MASK;
+
+  return EFI_SUCCESS;
+}
+
+/**
+  Get Partition info.
+  If This function is called to get partition descriptors
+  (Flags isn't set with FFA_PART_INFO_FLAG_TYPE_COUNT),
+  It should call ArmFfaLibRxRelease() to release RX buffer.
+
+  @param [in]   ServiceGuid    Service guid.
+  @param [in]   Flags          If this function called to get partition desc
+                               and get successfully,
+                               Caller should release RX buffer by calling
+                               ArmFfaLibRxRelease
+  @param [out]  Count          Number of partition or partition descriptor
+  @param [out]  Size           Size of Partition Info structure in Rx Buffer
+
+  @retval EFI_SUCCESS
+  @retval Others               Error
+**/
+EFI_STATUS
+EFIAPI
+ArmFfaLibPartitionInfoGet (
+  IN  EFI_GUID  *ServiceGuid,
+  IN  UINT32    Flags,
+  OUT UINT32    *Count,
+  OUT UINT32    *Size           OPTIONAL
+  )
+{
+  EFI_STATUS    Status;
+  ARM_FFA_ARGS  FfaArgs;
+  UINT32        SmcUuid[4];
+
+  if (Count == NULL) {
+    return EFI_INVALID_PARAMETER;
+  }
+
+  if ((((Flags >> FFA_PART_INFO_FLAG_TYPE_SHIFT) & FFA_PART_INFO_FLAG_TYPE_MASK) !=
+       FFA_PART_INFO_FLAG_TYPE_COUNT) && (Size == NULL))
+  {
+    return EFI_INVALID_PARAMETER;
+  }
+
+  if (ServiceGuid != NULL) {
+    CopyGuid ((EFI_GUID *)SmcUuid, ServiceGuid);

```

Comment:
I don't think so.
When you see the FF-A specification explaining UUID, 
It mention SMC Calling Convention v1.2. (See https://developer.arm.com/documentation/den0028/c),
and UUID describe in rfc4122's binary format comptible for EFI_GUID (only field name different).

So FF-A and EFI_GUID is comptible UUID, It doesn't need any translation.

Reply:
But somehow it seems that we have to perform such translation in order to have FFA look up the service/partition ID properly: https://gitlab.arm.com/arm-reference-solutions/edk2/-/blob/RD-INFRA-2022.12.22/ArmPkg/Drivers/MmCommunicationDxe/MmCommunication.c?ref_type=tags#L430. Can you please let me know why that is the case?

Reply:
Because It's way of saving UUID in TF-A manifest implementation.

for example for UUID  {378daedc-f06b-4446-831440ab933c87a3}

In TF-A, the UUID saved in each 32 bit integer.
    <0x378daedc 0xf06b4446 0x831440ab 0x933c87a3>

and TF-A raw memory:
    dc ae 8d 37
    46 44 6b f0
    ab 40 14 83
    a3 87 3c 93

But the EFI_GUID should be changed with this format. which saved like
    dc ae 8d 37
    6b f0 46 44
    83 14 40 ab
    93 3c 87 a3

But spec's uuid format should follow EFI_GUID. I think the TF-A's manifest defines UUID seems wrong.
So in manifest, for 378daedc-f06b-4446-831440ab933c87a3,
It should save as

<0x378daedc 0x4446f06b 0xab401483 0xa3873c93>

Reply:
Okay, that is good to know. An additional ask is would you be willing to update the spec/implementation for the manifest? :) With the current manifest implementation + this PR, I am afraid the service query will not work properly.